bentoml/_internal/io_descriptors/numpy.py

```diff
@@ -546,15 +359,176 @@ async def predict(input: NDArray[np.int16]) -> NDArray[Any]:
         """
         if isinstance(sample_input, np.generic):
             raise BentoMLException(
-                "NumpyNdarray.from_sample() expects a numpy.array, not numpy.generic."
+                "'NumpyNdarray.from_sample()' expects a 'numpy.array', not 'numpy.generic'."
             )
 
-        inst = cls(
+        klass = cls(
             dtype=sample_input.dtype,
             shape=sample_input.shape,
             enforce_dtype=enforce_dtype,
             enforce_shape=enforce_shape,
         )
-        inst.sample_input = sample_input
+        klass.sample_input = sample_input
+
+        return klass
 
-        return inst
+    async def from_grpc_request(
+        self, request: pb.Request, context: BentoServicerContext
+    ) -> ext.NpNDArray:
+        """
+        Process incoming protobuf request and convert it to ``numpy.ndarray``
+
+        Args:
+            request: Incoming RPC request message.
+            context: grpc.ServicerContext
+
+        Returns:
+            a ``numpy.ndarray`` object. This can then be used
+             inside users defined logics.
+        """
+        from ..utils.grpc import get_field
+        from ..utils.grpc import raise_grpc_exception
+        from ..utils.grpc import validate_content_type
+        from ..utils.grpc.mapping import dtypepb_to_fieldpb_map
+        from ..utils.grpc.mapping import dtypepb_to_npdtype_map
+        from ..utils.grpc.mapping import fieldpb_to_npdtype_map
+        from ..utils.grpc.mapping import npdtype_to_fieldpb_map
+
+        # validate gRPC content type if content type is specified
+        validate_content_type(context, self)
+
+        field = get_field(request, self)
+
+        # The logic behind processing shape is as follows:
+        # 1. If 'shape' is specified in the request message, then:
+        #    - if 'shape' is specified inside IO Descriptor, we will omit this value unless 'enforce_shape' is True.
+        #    - Otherwise, use 'shape' from message
+        #
+        # 2. If the 'shape' is not specified inside the request message, then:
+        #    - If 'shape' is not defined at the IO descriptor level, 'shape' value will be omitted unless 'enforce_shape' is True.
+        #    - Otherwise, we will use shape from the IO descriptor.
+        if self._enforce_shape:
+            logger.debug(
+                f"'enforce_shape={self._enforce_shape}', using '{self._shape}'..."
+            )
+            shape = self._shape
+        else:
+            shapepb = field.shape  # this can be an empty list ([])
+            if shapepb:
+                shape = tuple(shapepb)
+                if self._shape:
+                    logger.warning(
```

Comment:
When force_shape is True, we should always use the shape specified when creating the Numpy io descriptor, not using the shape in request. Let's just raise an exception if the specified shape is different from the shape in the request if enforce shape is True. Make sure to handle the `-1` case.

Reply:
This is checked on L410 right?